stripe/_http_client.py

```diff
@@ -334,6 +357,106 @@ def close(self):
+            method, url, headers, post_data, is_streaming=False, usage=_usage
+        )
+
+    async def request_stream_with_retries_async(
+        self,
+        method,
+        url,
+        headers,
+        post_data=None,
+        *,
+        _usage: Optional[List[str]] = None
+    ) -> Tuple[Any, int, Any]:
+        return await self._request_with_retries_internal_async(
+            method, url, headers, post_data, is_streaming=True, usage=_usage
+        )
+
+    @classmethod
+    async def sleep(cls: Type[Self], secs: float) -> Awaitable[None]:
+        raise NotImplementedError(
+            "HTTPClientAsync subclasses must implement `sleep`"
+        )
+
+    async def _request_with_retries_internal_async(
+        self, method, url, headers, post_data, is_streaming, usage
+    ):
+        self._add_telemetry_header(headers)
+
+        num_retries = 0
+
+        while True:
+            request_start = _now_ms()
+
+            try:
+                if is_streaming:
+                    response = await self.request_stream_async(
+                        method, url, headers, post_data
+                    )
+                else:
+                    response = await self.request_async(
+                        method, url, headers, post_data
+                    )
+                connection_error = None
+            except APIConnectionError as e:
+                connection_error = e
+                response = None
+
+            if self._should_retry(response, connection_error, num_retries):
+                if connection_error:
+                    _util.log_info(
+                        "Encountered a retryable error %s"
+                        % connection_error.user_message
+                    )
+                num_retries += 1
+                sleep_time = self._sleep_time_seconds(num_retries, response)
+                _util.log_info(
+                    (
+                        "Initiating retry %i for request %s %s after "
+                        "sleeping %.2f seconds."
+                        % (num_retries, method, url, sleep_time)
+                    )
+                )
+                await self.sleep(sleep_time)
+            else:
+                if response is not None:
+                    self._record_request_metrics(
+                        response, request_start, usage
+                    )
+
+                    return response
+                else:
+                    assert connection_error is not None
+                    raise connection_error
+
+    async def request_async(self, method, url, headers, post_data=None):
+        raise NotImplementedError(
+            "HTTPClientAsync subclasses must implement `request`"
+        )
+
+    async def request_stream_async(self, method, url, headers, post_data=None):
+        raise NotImplementedError(
+            "HTTPClientAsync subclasses must implement `request_stream`"
+        )
+
+    async def close(self):
```

Comment:
close_async?

Reply:
Is there a linter for this?

Reply:
We'd have to write ourselves. The only thing I can find is https://pypi.org/project/flake8-async/

Reply:
I can ticket